packages/components/rating/src/use-rating.ts

```diff
@@ -0,0 +1,341 @@
+   *    radioButtonWrapper: "radio-button-wrapper-classes",
+   *    description: "description-classes",
+   *    errorMessage: "error-message-classes",
+   * }} />
+   * ```
+   */
+  classNames?: SlotsToClasses<RatingSlots>;
+  /**
+   * onValueChnage method gets called whenever the value of rating changes.
+   */
+  onValueChange?: (value: number) => void;
+}
+
+export type UseRatingProps = Props & RatingVariantProps & Omit<AriaTextFieldProps, "onChange">;
+
+export function useRating(originalProps: UseRatingProps) {
+  const globalContext = useProviderContext();
+  const [props, variantProps] = mapPropsVariants(originalProps, rating.variantKeys);
+
+  const {
+    ref,
+    baseRef,
+    as,
+    className,
+    length = 5,
+    classNames,
+    strokeColor,
+    precision: precisionValue = 1,
+    fillColor = "gold",
+    opacity: opacityValue = 0.2,
+    activeOpacity: activeOpacityValue = 1,
+    children = null,
+    isSingleSelection = false,
+    icon,
+    defaultValue,
+    value,
+    onValueChange = () => {},
+    validationBehavior,
+    name,
+    errorMessage,
+    description,
+    onChange,
+    onBlur,
+    ...otherProps
+  } = props;
+
+  const {
+    disableAnimation = globalContext?.disableAnimation ?? false,
+    isDisabled = originalProps.isDisabled ?? false,
+    isRequired = originalProps.isRequired ?? false,
+  } = originalProps;
+
+  const {direction} = useLocale();
+  const isRTL = direction === "rtl";
+
+  const Component = as || "div";
+
+  const domRef = useDOMRef<HTMLInputElement>(ref);
+  const baseDomRef = useDOMRef<HTMLDivElement>(baseRef);
+  const iconWrapperRef = useRef<HTMLDivElement>(null);
+
+  const slots = useMemo(
+    () =>
+      rating({
+        ...variantProps,
+        disableAnimation,
+        isDisabled,
+      }),
+    [objectToDeps(variantProps)],
+  );
+
+  const handleValueChange = useCallback(
+    (value: RatingValueType | undefined) => {
+      if (value && value.selectedValue != undefined && value.selectedValue != -1) {
+        onValueChange(value.selectedValue);
+      }
+    },
+    [onValueChange],
+  );
+
+  const [ratingValue, setRatingValue] = useControlledState<RatingValueType>(
+    props.value
+      ? {hoveredValue: Number(props.value), selectedValue: Number(props.value)}
+      : undefined,
+    props.defaultValue
+      ? {hoveredValue: Number(props.defaultValue), selectedValue: Number(props.defaultValue)}
+      : {hoveredValue: -1, selectedValue: -1},
+    handleValueChange,
+  );
+
+  // if we use `react-hook-form`, it will set the input value using the ref in register
+  // i.e. setting ref.current.value to something which is uncontrolled
+  // hence, sync the state with `ref.current.value`
+  useSafeLayoutEffect(() => {
+    if (!domRef.current || !domRef.current.value) return;
+    setRatingValue({
+      hoveredValue: Number(domRef.current.value),
+      selectedValue: Number(domRef.current.value),
+    });
+  }, [domRef.current]);
```

Comment:
We do use this way whenever we have an internal hidden input.
A similar approach is being used in `use-input.ts` as well.

Reply:
We use similar approach in `use-input.ts` in Input Component.
In Radio group while using the react-hook-form as below:
```
const {
   register
} = useForm({
   defaultValues: {
       radio: "defaultValue"
   }
})
```
the defaultValue would not be visible on the radioGroup, but ref will get populated with the defaultValue. That is when we submit the form, the form data will have "defaultValue" for the field.

To avoid this, the above way is used here.